### PR TITLE
[SPARK-48791][CORE][3.4] Fix perf regression caused by the accumulators registration overhead using CopyOnWriteArrayList

### PR DESCRIPTION
**Pull request by Ngone51**
This PR backports https://github.com/apache/spark/pull/47197 to branch-3.4.

### What changes were proposed in this pull request?

This PR proposes to use the `ArrayBuffer` together with the read/write lock rather than `CopyOnWriteArrayList` for `TaskMetrics._externalAccums`.

### Why are the changes needed?

Fix the perf regression that caused by the accumulators registration overhead using `CopyOnWriteArrayList`.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Manually tested.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by LuciferYang — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @Ngone51 . Could you double-check the CI failure and make it happy? It seems that there is Scala 2.13 compilation failures in CIs.

```
[error] /home/runner/work/spark/spark/mllib-local/src/main/scala/org/apache/spark/ml/stat/distribution/MultivariateGaussian.scala:117:34: Symbol 'type scala.collection.compat.package.IterableOnce' is missing from the classpath.
[error] This symbol is required by 'type breeze.linalg.support.LowPrioCanTraverseValues.X'.
[error] Make sure that type IterableOnce is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
[error] A full rebuild may help if 'LowPrioCanTraverseValues.class' was compiled against an incompatible version of scala.collection.compat.package.
[error]     val tol = Utils.EPSILON * max(d) * d.length
[error]                                  ^
[error] one error found
```

**Comment by dongjoon-hyun** [comment]
This PR seems to be ready, but let's wait until we merge #47297 because we need to keep the backporting order `master` -> `branch-3.5` -> `branch-3.4` to prevent any future regression.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to branch-3.4.

**Comment by dongjoon-hyun** [comment]
Thank you, @Ngone51 and @LuciferYang .

**Pull request closed without merging**
